src/util/type.h

```diff
@@ -17,8 +20,37 @@ Author: Daniel Kroening, kroening@kroening.com
 
 class namespacet;
 
-/*! \brief The type of an expression
-*/
+/// The type of an expression, extends irept. Types may have subtypes. This is
+/// modeled with two subs named “subtype” (a single type) and “subtypes”
+/// (a vector of types). The class typet only adds specialized methods
+/// for accessing the subtype information to the interface of irept.
+/// Pre-defined types:
```

Comment:
Two requests here, although this may be in partial conflict with the documentation-only push taking place:
1) I think the list of pre-defined types should live in `std_types.h`
2) The functions `is_number` and (less clear) `is_constant_or_has_constant_components` should also move to `std_types.h`. It is `std_types.{h,cpp}` that actually declares the various "number" types.

Reply:
I can do the second point in a separate PR where I will be cleaning the code a bit too - adding `can_cast_type` to those types that don;t have it yet.

Reply:
I think this list can go away. It's incomplete and outdated.

Reply:
The list of pre-defined types seems vastly out of date to me. I would just remove it. Do you think it would be of use to have an exhaustive list of defined types in std_types.h file description?

Reply:
(But I'm still in favour of the follow-up PR moving `is_number` to a different file.)